### PR TITLE
DEVPROD-755 Improve check blocked tasks job and documentation to related code

### PR DESCRIPTION
**Pull request by ZackarySantana**
DEVPROD-755

### Description
The ticket was created to address a couple things:
1. If there are duplicate db queries
2. Add comments/documentation
3. Mark the task as not finished if there is trouble updating its dependencies

#### Solutions
1. (This all relates to the PR linked in the ticket specifically) I scrummed through the code and didn't find any duplicate calls. There could be duplicate calls for the container task queue because the only container tasks in the queue are already vetted to have their dependencies satisfied-- but after asking Kim for some context about it, she mentioned containers are in maintenance mode and it might be too extra optimizing it as such.

2. I added some comments and documentation, I also refactored some names to make it more self-documenting. I as well added some tests to prevent regression and make sure we are using the cache.

3. I'm not sure how this behavior should exactly look. It stems from [this comment](https://github.com/evergreen-ci/evergreen/pull/6900#discussion_r1320279512) on the previous work's PR. I simply added an error check that sees if getting or updating the blocked finished tasks errored, and if it did, mark the current task as not finished.

### Testing
CI testing

### CODE REVIEW

File: units/check_blocked_tasks.go [hunk 1]
```diff
@@ -135,21 +139,21 @@ func checkUnmarkedBlockingTasks(ctx context.Context, t *task.Task, dependencyCac
 		return nil
 	}
 
-	blockingTasks, err := t.RefreshBlockedDependencies(dependencyCaches)
+	finishedBlockingTasks, err := t.GetFinishedBlockingDependencies(dependencyCaches)
 	catcher.Wrap(err, "getting blocking tasks")
 	blockingTaskIds := []string{}
 	if err == nil {
-		for _, blockingTask := range blockingTasks {
+		for _, blockingTask := range finishedBlockingTasks {
```
Comment (hadjri) [inline units/check_blocked_tasks.go:146]: Re this [comment](https://github.com/evergreen-ci/evergreen/pull/6900#discussion_r1320279512): should we include an extra `MarkDependenciesFinished` call here to updates all direct dependencies on this task to cache whether or not this task has finished running

**Review by hadjri — CHANGES_REQUESTED** [review]

File: model/task_lifecycle.go [hunk 2]
```diff
@@ -906,21 +906,17 @@ func UpdateBlockedDependencies(ctx context.Context, t *task.Task) error {
 		return errors.Wrapf(err, "getting tasks depending on task '%s'", t.Id)
 	}
 
-	buildIDsSet := make(map[string]struct{})
+	buildIDs := []string{}
```
[the remark this answers is not in the document]
Comment (ZackarySantana) [reply model/task_lifecycle.go:909]: I'll switch it back over and document that it's to ensure no duplicates!

**Comment by ZackarySantana** [comment]
Im perfecting the query now- can't seem to get it down perfect but all the other logic should be stable to review- sorry about that!

File: model/task/db.go [hunk 3]
```diff
@@ -519,6 +519,23 @@ func FailedTasksByVersionAndBV(version string, variant string) bson.M {
 	}
 }
 
+func PotentiallyBlockedTasksByIds(taskIds []string) bson.M {
```
Comment (hadjri) [inline model/task/db.go:522]: Doc exported function

File: model/task/db_test.go [hunk 4]
```diff
@@ -229,6 +229,75 @@ func TestFailedTasksByVersion(t *testing.T) {
 	})
 }
 
+// use asserts and requires
```
Comment (hadjri) [inline model/task/db_test.go:232]: We can remove this comment

File: units/check_blocked_tasks.go [hunk 5]
```diff
@@ -99,7 +99,8 @@ func (j *checkBlockedTasksJob) getDistroTasksToCheck() []task.Task {
 		return nil
 	}
 
-	tasksToCheck, err := task.Find(task.ByIds(taskIds))
+	// if len(t.DependsOn) == 0 || t.OverrideDependencies || !utility.IsZeroTime(t.DependenciesMetTime) {
```
Comment (hadjri) [inline units/check_blocked_tasks.go:102]: Is this meant to be uncommented or removed?

File: model/task/task.go [hunk 6]
```diff
@@ -816,6 +805,21 @@ func (t *Task) BlockedOnDeactivatedDependency(depCache map[string]Task) ([]strin
 	return blockingDeps, nil
 }
 
+func populateDependencyTaskCacheSingular(depCache map[string]Task, depId string) (*Task, error) {
```
Comment (hadjri) [inline model/task/task.go:808]: Can we comment this because its purpose might not be clear at first glance

File: model/task/db.go [hunk 7]
```diff
@@ -519,6 +519,23 @@ func FailedTasksByVersionAndBV(version string, variant string) bson.M {
...
+	return bson.M{
+		IdKey: bson.M{"$in": taskIds},
+		"$and": []bson.M{
+			{DependsOnKey: bson.M{
+				"$exists": true,
+				"$not":    bson.M{"$size": 0},
+			}},
+			{OverrideDependenciesKey: false},
```
Comment (hadjri) [inline model/task/db.go:530]: OverrideDependenciesKey not existing would also be valid

**Review by hadjri — CHANGES_REQUESTED** [review]

**Review by hadjri — APPROVED** [review]

**Pull request merged**